### ISSUE TITLE
kv: triage microbenchmark regressions for v23.2

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#114310]
From the 23.2 microbenchmark results: https://docs.google.com/spreadsheets/d/1l7yX9btMitDyZ1DLJQDaJVNoukpYbk2iu1tgTg1551Q/edit#gid=5

All relevant performance regressions are (at least) reflected in the `sec/op` metric. Some are also reflected in heap allocation metrics, which may explain the slowdown.

I've done my best to group related regressions to avoid redundant work.

`pkg/kv/kvclient/kvcoord`

- [x] `ReturnOnRangeBoundary` @arulajmani @nvanbenschoten 
- [x] `TruncateLoop` (and also `TruncateLegacy`) @arulajmani 

`pkg/kv/kvserver`

- [x] `BumpSideTransportClosed` @arulajmani (https://github.com/cockroachdb/cockroach/issues/115058)
- [x] `MVCCGCWithForegroundTraffic` @kvoli 
- [x] `OptimisticEval` (and also `OptimisticEvalForLocks`) @arulajmani 
- [x] `ReplicaProposal`

`pkg/kv/kvserver/batcheval`

- [x] `RefreshRange` @miraradeva

`pkg/kv/kvserver/concurrency`

- [x] `LockTable` @arulajmani 

`pkg/kv/kvserver/spanlatch`

- [x] `LatchManagerReadWriteMix` @arulajmani 

`pkg/kv/kvserver/logstore` 

- [x] `LogStore_StoreEntries` @kvoli 

`pkg/kv/kvserver/raftentry`

- [x] `EntryCacheClearTo` @andrewbaptist 

`pkg/kv/kvserver/raftlog` 

- [x] `RaftAdmissionMetaOverhead` @nvanbenschoten 

`pkg/kv/kvserver/tscache`

- [x] `IntervalSklAdd` @nvanbenschoten 
- [x] `IntervalSklDecodeValue` @nvanbenschoten 
- [x] `TimestampCacheInsertion` @nvanbenschoten 

`pkg/kv/kvserver/concurrency` and `pkg/kv/kvserver/spanlatch`

These packages are seeing regressions across a collection of btree benchmarks. If this is not noise (it probably is), they likely have the same cause.

- [x] `BTreeDelete`, `BTreeDeleteInsert`, `BTreeInsert`, etc. @nvanbenschoten 
- [x] `BTreeIterOverlapScan` @nvanbenschoten 

----

### Tips for debugging

[Benchdiff](https://github.com/nvanbenschoten/benchdiff) can be a useful tool for debugging. Install the tool, then follow these steps to auto-bisect the regression.
```sh
# Setup steps:
export base_sha=f13c021b
export master_sha=6d4a8274
echo 'git clean -e benchdiff -e post-checkout.sh -e .bazelrc.user -e _bazel -e bin/.dev-status -e bin/dev-versions -dxff' >> post-checkout.sh
echo './dev generate' >> post-checkout.sh
echo './dev build short' >> post-checkout.sh

# Per-benchmark configuration.
export pkg_name='./pkg/kv/kvserver/spanlatch'
export test_name='BTreeDeleteInsert/count=16'
export threshold=0.2

# Single comparison between master and release-23.1:
benchdiff --new=${master_sha} --old=${base_sha} --post-checkout='bash post-checkout.sh' --run=${test_name} --count=10 ${pkg_name}
# OR auto-bisect regression:
rm -rf ./benchdiff/*
git bisect reset
git bisect start ${master_sha} ${base_sha}
git bisect run benchdiff --old=${base_sha} --post-checkout='bash post-checkout.sh' --run=${test_name} --count=10 --threshold=${threshold} ${pkg_name}
```

Jira issue: CRDB-33448

**Comment by miraradeva** [comment]
I bisected the `RefreshRange` regression to this [Pebble bump commit](https://github.com/cockroachdb/cockroach/commit/91fe543a1618ff2ae159c87d38c320781b673926), which includes these commits:
[58a05cd4](https://github.com/cockroachdb/pebble/commit/58a05cd4) ingest: add ability to ingest external (not local or shared) files
[49821f58](https://github.com/cockroachdb/pebble/commit/49821f58#diff-fb3af69fa3b741fd8d5bcadb389b5ac25906b3272e95c18690092d5a3678ee4e) db: use ordinary ingestion seqnums for shared sstables
[fd0cd8db](https://github.com/cockroachdb/pebble/commit/fd0cd8db) workflows: fix trim-prefix in code-cover-gen.yaml
[71f4d6c6](https://github.com/cockroachdb/pebble/commit/71f4d6c6) workflows: fix trim-prefix in code-cover-gen.yaml
[c055f188](https://github.com/cockroachdb/pebble/commit/c055f188) github: code coverage workflows

Nothing looks particularly suspicious to me but I don't have good context here.

**Comment by arulajmani** [comment]
After a fair bit of golang version fighting, I bisected `TruncateLoop` down to https://github.com/cockroachdb/cockroach/pull/107658. The regression is resulting from the extra field we added to the `Header` proto, which is intended.

**Comment by arulajmani** [comment]
I tracked down the majority of the regression we're seeing in `OptimisticEval` to https://github.com/cockroachdb/cockroach/commit/8e72fe0d4d018819ffaeabd3df201fa19d01d733, which was already fixed in https://github.com/cockroachdb/cockroach/pull/114437.

**Comment by pavelkalinnikov** [comment]
Team, heads up on fix #115147 - the bug could have impacted some microbenchmarks, particularly those in which proposals are sent concurrently.

**Comment by nvanbenschoten** [comment]
I've extracted the regression in `RefreshRange` into https://github.com/cockroachdb/cockroach/issues/115215 and `RaftAdmissionMetaOverhead` into https://github.com/cockroachdb/cockroach/issues/115216.

**Comment by kvoli** [comment]
Bisecting MVCCGCWithForegroundTraffic w/ 0.05 threshold pointed to 199b177e4e835a668ca339e485baf1ba6e7ad69f, which appears unrelated. Bisected between 5f4fe6e2d9c1c40e74ba9f0248063a5f5a3ec4d7 and c0f0059ca529a18d20736d83013cc46f21c1ae67. 

```
git bisect log                                                         # bad: [c0f0059ca529a18d20736d83013cc46f21c1ae67] Merge #110083
# good: [5f4fe6e2d9c1c40e74ba9f0248063a5f5a3ec4d7] roachtest: set 30m timeout for all disk stall roachtests
git bisect start 'c0f0059ca529a18d20736d83013cc46f21c1ae67' '5f4fe6e2d9c1c40e74ba9f0248063a5f5a3ec4d7'
# bad: [462fd410f95d0f06129bdda84b344d9b95441cc4] Merge #105015
git bisect bad 462fd410f95d0f06129bdda84b344d9b95441cc4
# bad: [50a0afce0fcd306a22051345e9f3a8c6d93e5928] Merge #100432 #102416
git bisect bad 50a0afce0fcd306a22051345e9f3a8c6d93e5928
# good: [e640de72f80af91beeff2256e8fb25c966c2b845] Merge #99423 #99987 #100565
git bisect good e640de72f80af91beeff2256e8fb25c966c2b845
# bad: [f24b893e5d825bc5fd7b91858426063590e7bdef] Merge #100270
git bisect bad f24b893e5d825bc5fd7b91858426063590e7bdef
# bad: [61946783bc69e4fe9dab0f7702e7a0ae550af84d] Merge #100925 #100993
git bisect bad 61946783bc69e4fe9dab0f7702e7a0ae550af84d
# good: [7d219406b9c0fc5079403840d55ba40f721512f9] Merge #99958 #100726 #100741 #100778 #100821
git bisect good 7d219406b9c0fc5079403840d55ba40f721512f9
# bad: [570ec001cb66c434a9f382bffca9e4d167dac85b] Merge #100773 #100842 #100898
git bisect bad 570ec001cb66c434a9f382bffca9e4d167dac85b
# good: [0f2db80943ae1c8977527aeb89be9e7fc0bd97b2] pgwire: update comments for `limitedCommandResult`
git bisect good 0f2db80943ae1c8977527aeb89be9e7fc0bd97b2
# bad: [22ab7edf761e16c09b70caeecb27bd332bfe88ee] Merge #99663 #99947 #100188 #100620
git bisect bad 22ab7edf761e16c09b70caeecb27bd332bfe88ee
# bad: [199b177e4e835a668ca339e485baf1ba6e7ad69f] sql: disable multiple active portals for TestDistSQLReceiverDrainsMeta.
git bisect bad 199b177e4e835a668ca339e485baf1ba6e7ad69f
# good: [dcbcca9468d0e25dd790ea880d6bb3dacb82f36c] sql: store retErr and retPayload for pausable portals
git bisect good dcbcca9468d0e25dd790ea880d6bb3dacb82f36c
# good: [3d20ce5a9729ac189ea252ab9e6682425b10e65f] sql: correctly set inner plan for leafTxn when resuming flow for portal
git bisect good 3d20ce5a9729ac189ea252ab9e6682425b10e65f
# good: [789c2cdc27f169ea2d22552dd9906b523b31f126] pgwire: add tests for multiple active portals
git bisect good 789c2cdc27f169ea2d22552dd9906b523b31f126
# first bad commit: [199b177e4e835a668ca339e485baf1ba6e7ad69f] sql: disable multiple active portals for TestDistSQLReceiverDrainsMeta.
```

**Comment by nvanbenschoten** [comment]
> ReturnOnRangeBoundary has a ~3% regression between the first commit after the bump to go1.20 and the last commit before the bump to go1.21. 

The spreadsheet indicates that `ReturnOnRangeBoundary` regressed by 28.45% between f13c021b and 6d4a8274. Are we not able to reproduce that? The regression is on the order of 90µs (310µs -> 400µs), which is of a large enough magnitude that it's likely more than just noise.

**Comment by arulajmani** [comment]
> Are we not able to reproduce that?

My bad, I missed that line in the spreadsheet. I hadn't run the benchmark between `[base_sha, go_19_last_sha]` before -- doing so, I see:

```
benchdiff --new=${go_1_19_last_sha} --old=${base_sha} --post-checkout='bash post-checkout.sh' --run=${test_name} --count=10 ${pkg_name}
....
  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/kv/kvclient/kvcoord -

name                      old time/op    new time/op    delta
ReturnOnRangeBoundary-24     417µs ± 2%     965µs ±12%  +131.36%  (p=0.000 n=10+10)

name                      old alloc/op   new alloc/op   delta
ReturnOnRangeBoundary-24    93.8kB ± 1%   123.7kB ± 5%   +31.85%  (p=0.000 n=10+10)

name                      old allocs/op  new allocs/op  delta
ReturnOnRangeBoundary-24       762 ± 0%      1133 ± 5%   +48.71%  (p=0.000 n=10+10)
```

Running a bisect now.

**Comment by arulajmani** [comment]
Bisected a big chunk to 3502ca2ef2dffd9f19c3bd7d7277324433ac43ad. Running benchdiff between this and the prior commit:

```
  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/kv/kvclient/kvcoord -

name                      old time/op    new time/op    delta
ReturnOnRangeBoundary-24     514µs ±13%     767µs ±10%  +49.15%  (p=0.000 n=10+10)

name                      old alloc/op   new alloc/op   delta
ReturnOnRangeBoundary-24    96.9kB ± 4%   111.7kB ± 3%  +15.26%  (p=0.000 n=10+10)

name                      old allocs/op  new allocs/op  delta
ReturnOnRangeBoundary-24       799 ± 4%       984 ± 7%  +23.09%  (p=0.000 n=10+10)  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/kv/kvclient/kvcoord -

name                      old time/op    new time/op    delta
ReturnOnRangeBoundary-24     514µs ±13%     767µs ±10%  +49.15%  (p=0.000 n=10+10)

name                      old alloc/op   new alloc/op   delta
ReturnOnRangeBoundary-24    96.9kB ± 4%   111.7kB ± 3%  +15.26%  (p=0.000 n=10+10)

name                      old allocs/op  new allocs/op  delta
ReturnOnRangeBoundary-24       799 ± 4%       984 ± 7%  +23.09%  (p=0.000 n=10+10)
```

That doesn't fully explain the regression we saw above. I'll run another 2 bisects on each of the halves of `[base_sha, 3502ca2ef2dffd9f19c3bd7d7277324433ac43ad]` and `[3502ca2ef2dffd9f19c3bd7d7277324433ac43ad, go_19_last_sha]`

**Comment by arulajmani** [comment]
@kvoli mentioned that he landed https://github.com/cockroachdb/cockroach/pull/113229, which resolved some cpu/memory regressions introduced by https://github.com/cockroachdb/cockroach/commit/3502ca2ef2dffd9f19c3bd7d7277324433ac43ad. That checks out with what I saw while bisecting this as well -- I saw the benchmark improve between `[go_1_21_sha, master]` significantly. 

@nvanbenschoten is taking over the rest of the investigation (thanks!).

**Comment by nvanbenschoten** [comment]
`ReturnOnRangeBoundary` was tricky to pin down because its performance regressed and later recovered a few times during the release. I believe I was able to find [the regression](https://github.com/cockroachdb/cockroach/issues/115058#issuecomment-1858677868) which never recovered. Interestingly, the cause is the same as the `BumpSideTransportClosed` regression. We pulled that out into https://github.com/cockroachdb/cockroach/issues/115058, but we haven't yet bottomed out on why that commit has this effect. I've re-opened that issue to track this regression as well.

We're down to 0 remaining regressions that haven't either been resolved or identified and extracted into separate issues. Nice teamwork!
